pkg/xds/generator/egress/generator.go

```diff
@@ -0,0 +1,189 @@
+// by multiple zoneegress generators.
+type Resources struct {
+	Listener *envoy_listeners.ListenerBuilder
+}
+
+type ResourceInfo struct {
+	Proxy *core_xds.Proxy
+	Mesh  *core_mesh.MeshResource
+
+	Listener         Listener
+	Resources        Resources
+	Destinations     []*Destination
+	EndpointMap      core_xds.EndpointMap
+	ExternalServices []*core_mesh.ExternalServiceResource
+	TrafficRoutes    []*core_mesh.TrafficRouteResource
+	ZoneIngresses    []*core_mesh.ZoneIngressResource
+}
+
+// ZoneEgressGenerator is responsible for generating xDS resources for
+// a single ZoneEgressHost.
+type ZoneEgressGenerator interface {
+	Generate(xds_context.Context, *ResourceInfo) (*core_xds.ResourceSet, error)
+}
+
+// Generator generates xDS resources for an entire ZoneEgress.
+type Generator struct {
+	Generators []ZoneEgressGenerator
+}
+
+// ResourceBuilder is an interface commonly implemented by complex Envoy
+// configuration element builders.
+// TODO (bartsmykla): DRY pkg/plugins/runtime/gateway/builder.go:11
+type ResourceBuilder interface {
+	Build() (envoy_common.NamedResource, error)
+}
+
+// BuildResourceSet is an adaptor that triggers the resource builder,
+// b, to build its resource. If the builder is successful, the result is
+// wrapped in a ResourceSet.
+// TODO (bartsmykla): DRY pkg/plugins/runtime/gateway/builder.go:20
+func BuildResourceSet(b ResourceBuilder) (*core_xds.ResourceSet, error) {
+	resource, err := b.Build()
+	if err != nil {
+		return nil, err
+	}
+
+	if resource.GetName() == "" {
+		return nil, errors.Errorf("anonymous resource %T", resource)
+	}
+
+	set := core_xds.NewResourceSet()
+	set.Add(&core_xds.Resource{
+		Name:     resource.GetName(),
+		Origin:   OriginEgress,
+		Resource: resource,
+	})
+
+	return set, nil
+}
+
+func makeListener(zoneEgress *core_mesh.ZoneEgressResource) Listener {
+	networking := zoneEgress.Spec.GetNetworking()
+
+	address := networking.GetAddress()
+	port := networking.GetPort()
+
+	return Listener{
+		Port:         port,
+		Address:      address,
+		ResourceName: envoy_names.GetInboundListenerName(address, port),
+	}
+}
+
+func routeDestinations(
+	proxy *core_xds.Proxy,
+	mesh *core_mesh.MeshResource,
+) []*Destination {
+	var destinations []*Destination
+
+	meshName := mesh.GetMeta().GetName()
+	zoneEgressProxy := proxy.ZoneEgressProxy
+	endpointMap := zoneEgressProxy.MeshEndpointMap[meshName]
+
+	for serviceName, endpoints := range endpointMap {
+		for _, endpoint := range endpoints {
+			tags := map[string]string{
+				mesh_proto.ServiceTag: serviceName,
+				"mesh":                meshName,
+			}
+
+			destinations = append(destinations, &Destination{
+				Destination:       tags,
+				Weight:            endpoint.Weight,
+				IsExternalService: endpoint.IsExternalService(),
+			})
+		}
+	}
+
+	return destinations
+}
+
+func (g Generator) Generate(
+	ctx xds_context.Context,
+	proxy *core_xds.Proxy,
+) (*core_xds.ResourceSet, error) {
```

Comment:
I find this pattern to be overkill for egress. I see that you took inspiration from Gateway, but I don't think it's a good fit here. It makes sense in Gateway because we generate a lot of resources that depends on each other and I think we will extend it more and more.

Here I think it's pretty static, generators depends on listener and that's it.
I don't think we will add more Egress specific generator so I feel like building those interfaces `BuildResourceSet` etc. is all redundant. 

What I would do here is
* Create a ListenerBuilder in this function (inline listener_builder.go)
* For every mesh call two generators by passing `proxy`, `listener` and `mesh related resources` add generated resources to resource set
* Build a listener

Reply:
addressed